### PR TITLE
Fix native engine bug in editor

### PR DESCRIPTION
**Pull request by dogeFu**
Re: #

### Changelog

*

-------

### Continuous Integration

This pull request:

* [ ] needs automatic test cases check.
  > Manual trigger with `@cocos-robot run test cases` afterward.
* [ ] does not change any runtime related code or build configuration
  > If any reviewer thinks the CI checks are needed, please uncheck this option, then close and reopen the issue.

-------

### Compatibility Check

This pull request:

* [ ] changes public API, and have ensured backward compatibility with [deprecated features](https://github.com/cocos/cocos-engine/blob/v3.5.0/docs/contribution/deprecated-features.md).
* [ ] affects platform compatibility, e.g. system version, browser version, platform sdk version, platform toolchain, language version, hardware compatibility etc.
* [ ] affects file structure of the build package or build configuration which requires user project upgrade.
* [ ] **introduces breaking changes**, please list all changes, affected features and the scope of violation.

### CODE REVIEW

**Review by PatriceJiang — APPROVED** [review]

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp [hunk 1]
```diff
@@ -61,6 +61,18 @@ ccstd::unordered_map<ccstd::string, unsigned> jsbFunctionInvokedRecords;
         if (!(cond)) return val
 
 namespace se {
+AutoHandleScope::AutoHandleScope()
+: _handleScope(v8::Isolate::GetCurrent()) {
+#if CC_EDITOR
+    ScriptEngine::getInstance()->_getContext()->Enter();
+#endif
```
Comment (PatriceJiang) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: This would change the defination of HandleScope

Comment (dumganhar) [reply native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: The reason of this is to make sure the `context` is the one which we initializing engine. Cocos Creator Editor depends on Electron JS environment which may have more than one JS `context`. And while `Inspector Dev Tool` is attached, the current context may be a different one that will cause crash.

Yep, I agree that it seems that we modify the `Handle Scope` behavior a little bit, but I think it's currently the best place to fix that crash issue and it's wrapped in `CC_EDITOR` macro that will not affect `runtime` mode. Also binding code doesn't need to be changed. Besides, engine only uses one `context` now.

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp [hunk 2]
```diff
@@ -502,13 +514,14 @@ ScriptEngine::~ScriptEngine() { //NOLINT(bugprone-exception-escape)
...
     v8::HandleScope hs(_isolate);
+#if !CC_EDITOR
     _isolate->Enter();
     _isolate->SetCaptureStackTraceForUncaughtExceptions(true, JSB_STACK_FRAME_LIMIT, v8::StackTrace::kOverview);
     _isolate->SetFatalErrorHandler(onFatalErrorCallback);
     _isolate->SetOOMErrorHandler(onOOMErrorCallback);
     _isolate->AddMessageListener(onMessageCallback);
     _isolate->SetPromiseRejectCallback(onPromiseRejectCallback);
-
+#endif
```
Comment (minggo) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: - please add some comment
- can create a function to do it?

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp [hunk 3]
```diff
@@ -61,6 +61,18 @@ ccstd::unordered_map<ccstd::string, unsigned> jsbFunctionInvokedRecords;
...
 
 namespace se {
+AutoHandleScope::AutoHandleScope()
+: _handleScope(v8::Isolate::GetCurrent()) {
+#if CC_EDITOR
+    ScriptEngine::getInstance()->_getContext()->Enter();
+#endif
+}
+
+AutoHandleScope::~AutoHandleScope() {
```
Comment (minggo) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: Please fix clang tidy error.

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp [hunk 4]
```diff
@@ -517,7 +530,7 @@ bool ScriptEngine::postInit() {
     _globalObj = Object::_createJSObject(nullptr, _isolate->GetCurrentContext()->Global());
     _globalObj->root();
     _globalObj->setProperty("window", Value(_globalObj));
-
+#if !CC_EDITOR
```
Comment (minggo) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: The same as above.

File: cocos/core/math/math-native-ext.ts [hunk 5]
```diff
@@ -105,4 +106,12 @@ if (NATIVE) {
...
     Object.setPrototypeOf(jsb.Color.prototype, Color.prototype);
     Object.defineProperty(jsb.Color.prototype, 'type', { configurable: true, enumerable: true, writable: false, value: MathType.COLOR });
+
+    ccclass('cc.Vec4')(jsb.Vec4);
+    ccclass('cc.Vec3')(jsb.Vec3);
+    ccclass('cc.Vec2')(jsb.Vec2);
+    ccclass('cc.Mat4')(jsb.Mat4);
+    ccclass('cc.Mat3')(jsb.Mat3);
+    ccclass('cc.Color')(jsb.Color);
+    ccclass('cc.Quat')(jsb.Quat);
```
Comment (minggo) [inline cocos/core/math/math-native-ext.ts:116]: Why need these code?

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp (hunk 3 above)
Comment (dumganhar) [inline native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: Add `//NOLINT` to ignore  the `lint` error.

File: native/cocos/renderer/GFXDeviceManager.h [hunk 6]
```diff
@@ -103,6 +103,9 @@ class CC_DLL DeviceManager final {
         if (tryCreate<GLES2Device>(info, &device)) return device;
 #endif
 
+#ifdef CC_EDITOR
+    Device::isSupportDetachDeviceThread = false;
```
Comment (dumganhar) [inline native/cocos/renderer/GFXDeviceManager.h]: Fix the indentation.

File: cocos/core/math/math-native-ext.ts (hunk 5 above)
Comment (dogeFu) [reply cocos/core/math/math-native-ext.ts:116]: Editor uses 'cc.Vec4' to identify the type,if these decorator is missing, Editor will fail to recognize the type, resulting in a serialization error

Comment (PatriceJiang) [reply cocos/core/math/math-native-ext.ts:116]: The deserialization procedure needs to get the type information

File: native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp (hunk 3 above)
Comment (dogeFu) [reply native/cocos/bindings/jswrapper/v8/ScriptEngine.cpp]: done

**Review by minggo — APPROVED** [review]

**Review by dumganhar — APPROVED** [review]

**Pull request merged**